File: backend/src/guitar_player/schemas/analytics.py
"""Schemas for analytics tracking and dashboard responses."""

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class TrackEventRequest(BaseModel):
    event_type: str = Field(min_length=1, max_length=50)
    event_category: str = Field(min_length=1, max_length=30)
    song_id: uuid.UUID | None = None
    song_title: str | None = Field(default=None, max_length=500)
    session_id: str | None = Field(default=None, max_length=64)
    properties: dict[str, Any] | None = None


class TrackEventsRequest(BaseModel):
    events: list[TrackEventRequest] = Field(default_factory=list, max_length=50)


class TrackEventsResponse(BaseModel):
    accepted: int


class AnalyticsAccessResponse(BaseModel):
    allowed: bool
    email: str | None = None


class AnalyticsOverview(BaseModel):
    total_events: int
    unique_users: int
    total_sessions: int
    login_count: int
    song_play_count: int


class TimeBucket(BaseModel):
    bucket_start: datetime
    count: int


class EventTrend(BaseModel):
    event_type: str
    buckets: list[TimeBucket]


class SongRanking(BaseModel):
    song_id: uuid.UUID | None = None
    song_title: str | None = None
    play_count: int
    unique_users: int


class UserActivity(BaseModel):
    user_email: str
    event_count: int
    last_seen_at: datetime


class EventTypeBreakdown(BaseModel):
    event_type: str
    count: int


class RecentEvent(BaseModel):
    id: uuid.UUID
    created_at: datetime
    event_type: str
    event_category: str
    event_source: str
    user_email: str | None = None
    tenant_id: str | None = None
    aws_account_id: str | None = None
    song_id: uuid.UUID | None = None
    song_title: str | None = None
    session_id: str | None = None
    properties: dict[str, Any] | None = None


class UserEmailListResponse(BaseModel):
    items: list[str]


class AnalyticsDashboard(BaseModel):
    window_start: datetime
    window_end: datetime
    granularity: Literal["day", "week", "month"]
    overview: AnalyticsOverview
    trends: list[EventTrend]
    event_breakdown: list[EventTypeBreakdown]
    top_songs: list[SongRanking]
    user_activity: list[UserActivity]
    recent_events: list[RecentEvent]

File: frontend/src/types/analytics.ts
export interface AnalyticsOverview {
    total_events: number
    unique_users: number
    total_sessions: number
    login_count: number
    song_play_count: number
}

export interface TimeBucket {
    bucket_start: string
    count: number
}

export interface EventTrend {
    event_type: string
    buckets: TimeBucket[]
}

export interface SongRanking {
    song_id: string | null
    song_title: string | null
    play_count: number
    unique_users: number
}

export interface UserActivity {
    user_email: string
    event_count: number
    last_seen_at: string
}

export interface EventTypeBreakdown {
    event_type: string
    count: number
}

export interface RecentEvent {
    id: string
    created_at: string
    event_type: string
    event_category: string
    event_source: string
    user_email: string | null
    tenant_id: string | null
    aws_account_id: string | null
    song_id: string | null
    song_title: string | null
    session_id: string | null
    properties: Record<string, unknown> | null
}

export interface AnalyticsDashboard {
    window_start: string
    window_end: string
    granularity: 'day' | 'week' | 'month'
    overview: AnalyticsOverview
    trends: EventTrend[]
    event_breakdown: EventTypeBreakdown[]
    top_songs: SongRanking[]
    user_activity: UserActivity[]
    recent_events: RecentEvent[]
}

export interface TrackingEvent {
    event_type: string
    event_category: string
    song_id?: string | null
    song_title?: string | null
    session_id?: string | null
    properties?: Record<string, unknown>
}

export interface TrackEventsRequest {
    events: TrackingEvent[]
}

export interface TrackEventsResponse {
    accepted: number
}

export interface AnalyticsAccessResponse {
    allowed: boolean
    email: string | null
}

export interface AnalyticsUserEmailsResponse {
    items: string[]
}

export interface AnalyticsQueryParams {
    since?: string
    until?: string
    days?: number
    granularity?: 'day' | 'week' | 'month'
    user_email?: string
    event_type?: string
    limit?: number
}
